Docs (#129)
datajoint/datajoint-labbook: docs/sphinx/dev-docs.rst

 
Table List Component
--------------------
Table List view is responsible for the following:
- After the user selected a schema, fetch all the tables and its type from the backend
- Display all the tables and restrict if the user types a non empty string into the search box
- Upon the selection of a table, use the callback to send the table info back to the SideMenu component where it will send the info back to the home component.
 
Table View Component
====================
.. image:: _static/images/TableViewHighlight.png
 
Responsible for the handling of two sub components, the TableContent and TableInfo, along with fetching of the data required by both of them once 
this component receive a valid table name
 
Table Content
-------------
 
TODO: Need to add image here
Table Content is mainly responsible for the viewing, filtering, insertion, update, and deletion of tuples given a table, as such it is divided into 5 components:
- Table Content View (Display the tuples of the table)
- Filter Component (Filter the tuples of the table)
- Insert Component (Single tuple at a time)
- Update Component (Single tuple at a time)

> - Deletion Component (Single tuple at a time, uses delete_quick on the backend thus the tuple must not have any dependencies)

Reviewer: ```suggestion
- Deletion Component (Delete a single record at a time. Currently, the record must not have any child dependencies.)
```

 
Table Content View
~~~~~~~~~~~~~~~~~~
Table Content View is responsible for:

> - Fetching and display tuple

Reviewer: ```suggestion
- Fetch and display a record
```

> - Contain the Filter, Insert, Update, and Delete

Reviewer: ```suggestion
- Contain the ``Filter``, ``Insert``, ``Update``, and ``Delete`` components.
```

> - Allowing the user to check a specific tuple to send one of the specific subcomponent such as insert, update, and delete_quick

Reviewer: ```suggestion
- Allow user to check a specific record to send one of the specific subcomponent (such as insert, update, and delete).
```

 
Filter Component
~~~~~~~~~~~~~~~~
.. image:: _static/images/FilterComponentHighlight.png
Filter Component Notes:

> - Allow the user to filter the tuples based on the available attributes (Cannot filter by blob or long blob)

Reviewer: ```suggestion
- Allow the user to filter the records based on the available attributes (Currently, filtering by blob or long blob is not supported).
```

> - Upon each change, all the restrictions (represented by filter cards) are checked to see if they are valid, if so then trigger fetch with the new restrictions

Reviewer: ```suggestion
- Upon each change, all the restrictions (represented by filter cards) are checked to see if they are valid. If so, a fetch is performed with the new restrictions.
```

> - There is 1 second delay from the last change (such as typing) before the back end get queried and the view updates

Reviewer: ```suggestion
- There is a 1-second delay from the last change (such as typing) before the back end is queried and the view is updated.
```

 
Insert Component
~~~~~~~~~~~~~~~~
.. image:: _static/images/InsertComponentHighlight.png
Insert Components Notes:

> - Allow the users to insert tuple using html inputs elements

Reviewer: ```suggestion
- Allow the users to insert a record using HTML inputs elements.
```

> - If there is a checked tuple in the content viewer than ask the user if they want to copy overview

Reviewer: ```suggestion
- If there is a checked record in the content viewer, it will ask the user if they want to copy over the record to the view.
```

 
Update Component
~~~~~~~~~~~~~~~~
.. image:: _static/images/UpdateComponentHighlight.png
Notes:

> - User must select a tuple from table content viewer which will be copy over to this component for update

Reviewer: ```suggestion
- User must select a record from the table content viewer which will be copy over to this component to perform the update.
```

> - Only allow update of none primary and none blob fields

Reviewer: ```suggestion
- Only allows updates of non-primary and non-blob fields.
```

 
Delete Component
~~~~~~~~~~~~~~~~
.. image:: _static/images/DeleteComponentHighlight.png
Notes:

> - User must select a tuple from a table content viewer which will be copied over to this component for deletion.

Reviewer: ```suggestion
- User must select a record from a table content viewer which will be copied over to this component for deletion.
```

> - Delete is using delete_quick which means the tuple cannot have any child dependencies otherwise delete will fail.

Reviewer: ```suggestion
- Currently, ``Delete`` uses a quick delete operation (you may find more details on this in the ``datajoint`` pip package in ``datajoint.table.Table.delete_quick``). This means the record cannot have any child dependencies otherwise the delete will fail and return a message with additional details.
```

=======================
Developer Documentation
=======================
 
 meant to give an overview of the design behind DJ-LabBook for developers who are interested in contributing to the code base.
 
The code is built using ``react-cra`` using ``typescript`` with the backend being built on ``flask`` and ``datajoint``. Thus, it is highly recommended to get familiar with React's components and typical architecture design.
 
Requirements:
=============
`Docker <https://www.docker.com/>`_
 
`Docker Compose <https://docs.docker.com/compose/>`_

Building the docs
=================

> Docs are based off sphinx and typedocs where the latter is used for autodoc-ing the typescript code base

Reviewer 2: ```suggestion
Docs are based on sphinx and typedocs where the latter is used for autodoc-ing the typescript code base
```
